File: assignments/a07/luhn.py
def digits_of(number):
    return [int(i) for i in str(number)]

def luhn_checksum(card_number):
    digits = digits_of(card_number)
    odd_digits = digits[-1::-2]
    even_digits = digits[-2::-2]
    print odd_digits, even_digits
    total = sum(odd_digits)
    for digit in even_digits:
        print 'adding',sum(digits_of(2 * digit))
        total += sum(digits_of(2 * digit))
    return total % 10

def is_luhn_valid(card_number):
    return luhn_checksum(card_number) == 0

File: solutions/a07-signup/scripts/Luhn.js
// JavaScript File

// Code adapted from https://en.wikipedia.org/wiki/Luhn_algorithm

var Luhn = (function () {

    function digits_of(number,step,init) {
        number = number.toString();
        var len = number.length;
        var digits = [];
        for( var i = init; i < len; i += step ) {
            digits.push(parseInt(number.charAt(len-i-1),10));
        }
        return digits;
    }
    
    function plus(x,y) {
        return x+y;
    }
    
    // See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce
    function sum(array) {
        return array.reduce(plus,0);
    }
    
    function test_sum() {
        if( sum([1,2,3,4]) != 10) {
            console.log("sum on 1,2,3,4 is not correct");
        } else {
            console.log("sum on 1,2,3,4 is correct");
        }
    }

    function checksum(card_number) {
        var odd_digits = digits_of(card_number,2,0);
        var even_digits = digits_of(card_number,2,1);
        var total = sum(odd_digits);
        for ( var i in even_digits) {
            var digit = even_digits[i];
            total += sum(digits_of(2 * digit, 1, 0));
        }
        return total % 10;
    }   

    function is_valid(card_number) {
        return checksum(card_number) == 0;
    }
    
    function test() {
        console.log('should be true:  '+is_valid('79927398713'));
        console.log('should be false: '+is_valid('79927398714'));
    }
    
    // here are the exported values.
    return { checksum: checksum, is_valid: is_valid, sum: sum };
})();
